### ISSUE TITLE
Channel shutdown should trigger pending calls to fail immediately instead of taking a few minutes

### ISSUE THREAD

**Issue by muxi** [grpc/grpc#14220]
When calling grpc_channel_destroy() on a channel and the channel state becomes SHUTDOWN, all calls that are still pending on this channel should fail.

This is currently not the case, reproduced by [Guggyco/firebase-network-bug-reproduce](https://www.google.com/url?q=https://github.com/Guggyco/firebase-network-bug-reproduce&sa=D&usg=AFQjCNE_2GB2dO-KDKxvks8ekfnADaDFWQ).

Debugging shows that the call's pending op RECV_STATUS_ON_CLIENT was not returned by completion queue. In the transport, function [`end_all_the_calls`](https://github.com/grpc/grpc/blob/123547c9625c56fdf5cb4ddd1df55ae0c785fa60/src/core/ext/transport/chttp2/transport/chttp2_transport.cc#L2324) was never called.

**Comment by AspirinSJL** [comment]
So when WiFi network disconnects on iOS, we call `grpc_channel_destroy()`, which calls `start_transport_op()` with a `disconnect_with_error`. 

The expectation is that we reach https://github.com/grpc/grpc/blob/master/src/core/ext/transport/chttp2/transport/chttp2_transport.cc#L1761 and eventually `end_all_the_calls()`.

But we never go through that path. Instead, `start_transport_op()` in the client channel will unref `disconnect_with_error` at https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/client_channel.cc#L667

@markdroth I'm wondering whether /how `disconnect_with_error` is passed down and trigger `end_all_the_calls()`.

CC @y-zeng

**Comment by markdroth** [comment]
@AspirinSJL, you're right, it does seem a bit weird that we don't wind up propagating the `disconnect_with_error` transport op down to each subchannel.  However, I don't think it would be safe to change the client_channel code to propagate that transport op down to all subchannels.  And even without us doing that, I think this bug should not be occurring anyway.

Keep in mind that we share subchannels between parent channels.  If you have two channels sharing a given subchannel, and one of those channels gets shut down, we don't actually want to shut down the subchannel, because it's still in use by the other channel.  So this means that it's not actually safe for the client_channel code to send the `disconnect_with_error` transport op down to every subchannel, because it does not know which subchannels may also be in use by another channel.

What the client_channel code *does* do is to shut down the LB policy, which causes the LB policy to unref all of the subchannels.  That unref actually gets translated to unreffing the channel stack of the subchannel:

https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/subchannel.cc#L748

For any subchannel for which there is no other channel holding a ref, this should cause us to unref the transport.  And when the transport gets unreffed, it should cancel any active streams.  If the transport is not actually returning recv_trailing_metadata for active streams when it gets unreffed, then the bug probably lies in the transport.

One or more of the following folks may be able to help understand this: @vjpai @ncteisen @yang-g 

Also, one word of advice: Please start by making sure we have a test that reproduces this problem.  Without that, we cannot reliably identify the source of the bug or prove that any potential fix actually works.

**Comment by AspirinSJL** [comment]
I see. I can understand the expected behavior now.

@muxi Which version is the bug reporter using? The subchannel file they are using looks quite different from the one on our master.

**Comment by markdroth** [comment]
I think there may be a problem in the case where there are two channels open to the same server and only one of them shuts down.  I've described that situation (and a possible but not ideal fix) in #14292.

I'm not sure this is actually what's happening in this case, because even if there was a second channel open to the same server, I would expect both of them to be destroyed at the same time by the same iOS hook, thus causing the subchannel to shut down and the transport to cancel all calls.  So the problem I just described may be unrelated to this bug, but I thought I'd mention it here for completeness.

As I said earlier, I think we need to start by finding a way to reproduce this, so that we can see what's actually happening.

**Comment by AspirinSJL** [comment]
This bug is very easy to reproduce by hand. By checking the refcount log, I see that the last pending ref of the subchannel is for `state_watcher`. And I've verified that the unref doesn't happen because the callback watching the connectivity state is never invoked. This seems to make sense, because the user is expecting to destroy the whole channel top-down. The user watches the network connectivity and calls  `grpc_channel_destroy()` when the network dies. Everything else has gone, but the "state_watcher" of the subchannel insists that it hasn't seen any connectivity change so it can't unref the subchannel.

@markdroth  By reproducing this, do you mean reproducing by test?

**Comment by AspirinSJL** [comment]
After closer investigating, I think there is something like circular dependency between connected_subchannel and subchannel_call, if we want to shut down the channel and the calls in a top-down style.

Since the user is using 1.8.4 version (although this bug also happens in 1.9.x), let's look[ the file in that version](https://github.com/grpc/grpc/blob/v1.8.4/src/core/ext/filters/client_channel/subchannel.cc).

Here are two paths to unref a connected_subchannel:
1. `grpc_subchannel_unref()` -> `disconnect()` -> `GRPC_CONNECTED_SUBCHANNEL_UNREF()`.
2. `subchannel_call_destroy()` always calls `GRPC_CONNECTED_SUBCHANNEL_UNREF()`.

From path 2 we can see that each subchannel_call holds a ref to the connected_subchannel, then it seems impossible to cancel the subchannel_call by destroying the connected_subchannel first (i.e., the first path). 

A workaround is to cancel the call directly when the user detects network disconnection instead of propagating the unrefs down. The following log shows that the subchannel_call unrefs the connected_subchannel after it is cancelled directly. Before this workaround is added, the refs to the connected_subchannel can only be decreased to 4.

```

2018-02-20 16:03:41.262 firebase-network-bug-reproduce[2250:761125] error:Error Domain=io.grpc Code=1 "Cancelled" UserInfo={NSLocalizedDescription=Cancelled}
trailers:{
}

xxxxxxxxxxxxxxxxxxxxxxxxx Directly cancelled the first call.

D0220 16:03:41.263512000 7000006656 transport.cc:41]                   CALL_STACK 0x12f0a8dc8:0x12f0a8568   REF 3->4 termination
D0220 16:03:41.263592000 7000006656 transport.cc:41]                   CALL_STACK 0x12f09b948:0x12f09b938   REF 2->3 perform_stream_op
D0220 16:03:41.263718000 7000006656 transport.cc:56]                   CALL_STACK 0x12f09b948:0x12f09b938 UNREF 3->2 chttp2
D0220 16:03:41.263755000 7000006656 transport.cc:56]                   CALL_STACK 0x12f09b948:0x12f09b938 UNREF 2->1 perform_stream_op
D0220 16:03:41.263862000 7000006656 transport.cc:56]                   CALL_STACK 0x12f0a8dc8:0x12f0a8568 UNREF 4->3 termination
2018-02-20 16:03:41.264 firebase-network-bug-reproduce[2250:761068] cancelled call
D0220 16:03:41.266058000 7000006656 transport.cc:41]                   CALL_STACK 0x12f0ab3c8:0x12f0aab68   REF 1->2 termination
D0220 16:03:41.268005000 6143422464 transport.cc:56]                   CALL_STACK 0x12f0a8dc8:0x12f0a8568 UNREF 3->2 completion
D0220 16:03:41.270660000 7000006656 transport.cc:56]                   CALL_STACK 0x12f0ab3c8:0x12f0aab68 UNREF 2->1 termination
2018-02-20 16:03:41.270 firebase-network-bug-reproduce[2250:761068] cancelled call
D0220 16:03:41.274642000 7000006656 transport.cc:41]                   CALL_STACK 0x12f0a8dc8:0x12f0a8568   REF 3->4 termination
D0220 16:03:41.274709000 7000006656 transport.cc:56]                   CALL_STACK 0x12f0a8dc8:0x12f0a8568 UNREF 4->3 termination
2018-02-20 16:03:41.274 firebase-network-bug-reproduce[2250:761068] cancelled call
D0220 16:03:41.275143000 6143422464 transport.cc:56]                   CALL_STACK 0x12f0a8dc8:0x12f0a8568 UNREF 2->1 completion
2018-02-20 16:03:43.522 firebase-network-bug-reproduce[2250:761125] error:Error Domain=io.grpc Code=1 "Cancelled" UserInfo={NSLocalizedDescription=Cancelled}
trailers:{
}

xxxxxxxxxxxxxxxxxxxxxxxxx Directly cancelled the second call.

D0220 16:03:43.523257000 6144569344 transport.cc:56]                   CALL_STACK 0x12f0ab3c8:0x12f0aab68 UNREF 1->0 destroy
D0220 16:03:43.523355000 6144569344 transport.cc:56]                   CALL_STACK 0x12f071548:0x12f071538 UNREF 1->0 client_channel_destroy_call
D0220 16:03:45.092736000 6144569344 transport.cc:56]                   CLIENT_SUBCHANNEL 0x1304a8cf0:0x1304a8cf0 UNREF 4->3 picked
D0220 16:03:48.166225000 6144569344 transport.cc:56]                   CLIENT_SUBCHANNEL 0x1304a8cf0:0x1304a8cf0 UNREF 3->2 subchannel_call

xxxxxxxxxxxxxxxxxxxxxxxxx The first call unreffed the connected_subchannel.

D0220 16:03:48.166369000 6144569344 transport.cc:56]                   CLIENT_CHANNEL 0x12eda4810:0x12eda4780 UNREF 2->1 call
D0220 16:03:48.168065000 6144569344 transport.cc:56]                   CALL_STACK 0x12f0a8dc8:0x12f0a8568 UNREF 1->0 destroy
D0220 16:03:48.168146000 6144569344 transport.cc:56]                   CALL_STACK 0x12f09b948:0x12f09b938 UNREF 1->0 client_channel_destroy_call
D0220 16:03:49.055567000 6144569344 transport.cc:56]                   CLIENT_SUBCHANNEL 0x1304a8cf0:0x1304a8cf0 UNREF 2->1 picked
D0220 16:03:50.034636000 6144569344 transport.cc:56]                   CLIENT_SUBCHANNEL 0x1304a8cf0:0x1304a8cf0 UNREF 1->0 subchannel_call

xxxxxxxxxxxxxxxxxxxxxxxxx The second call unreffed the connected_subchannel.

D0220 16:03:50.034821000 6144569344 transport.cc:56]                   CLIENT_CHANNEL 0x12eda4810:0x12eda4780 UNREF 1->0 call
D0220 16:03:55.332504000 6144569344 subchannel.cc:572]                 SUBCHANNEL: 0x12ee90930   WEAK_UNREF 0x1 -> 0x0 [state_watcher]
```

I think we do want to let subchannel_call hold a ref to the connected_subchannel.

If this rationale sounds reasonable, I think we'd better tell the iOS users to use the workaround and document this somewhere.

**Comment by srini100** [comment]
Hypothetically, if there is a separate channel refcnt to a subchannel, would it tell us that subchannel is not in use any more when that count hits zero and hence all calls on that subchannel can be destroyed?

I am not clear on the workaround here. Users need to also cancel the calls?

**Comment by AspirinSJL** [comment]
We currently have separate counts for external (strong) refs and internal (weak) refs. But we planned to remove this kind of refcounting, because it's complicated and the only use case is for subchannel.

When the users detect some network failure, they will destroy the channel hoping that the calls will also be cancelled. But actually the calls are not being cancelled, probably because of the refcounting pattern we use. A workaround can be explicitly cancelling the calls when the network failure is detected.

**Comment by markdroth** [comment]
@AspirinSJL, yes, I am suggesting that we have a test for this case.  Otherwise, we have no way to prove that any potential fix actually works or that we don't break that fix with subsequent changes in the future.

I don't think it's feasible to have the iOS code cancel all of the calls, because it does now have a list of all of the calls.  We don't currently keep a list of all calls on a channel anywhere.  If we actually need to cancel all of the calls (and I'm still not convinced that we do; see below), then it probably makes more sense to do this inside of core than to push this work onto the application or the iOS interface.

Keep in mind that there is not actually any requirement that the destruction happen on a top-down basis.  Or, to say this another way, we do not need to ensure that the calls get cancelled as a result of `grpc_channel_destroy()` being called from the surface.  As long as the calls will eventually finish, and as long as the channel gets destroyed when the last call does finish, I think that's fine.

I think the bug here is not that we are not destroying the calls when `grpc_channel_destroy()` is called; it's that a call started before `grpc_channel_destroy()` was called is not finishing afterwards.  So the question is, why is that not happening?

**Comment by srini100** [comment]
The bug here is that when grpc_channel_destroy() is called, the user is expecting all pending calls on that channel to return immediately. I think @AspirinSJL is suggesting the user to explicitly cancel calls along with destroying the channel. This mean the user has to keep track of calls on that channel. Just to be clear, the calls do return eventually after a long delay. The user is expecting the calls to return immediately when the channel is destroyed.

**Comment by markdroth** [comment]
Ah, okay, so the original description of this bug is wrong?  @muxi said that the RECV_STATUS_ON_CLIENT op was never returned, but it sounds like you're now saying that it is actually returned; the issue is actually that it's not being returned *immediately*, which is what's being asked for here.  Is that right?

If so, then I think this will require fixing #14292.

**Comment by muxi** [comment]
Sorry I was wrong on this. It seems the op returns after a long (a few minutes) delay.

From its description I think #14292 is the right direction.

**Comment by AspirinSJL** [comment]
Actually the workaround is not too bad to me. In iOS layer, we have two kinds of network monitor: a single channel-level monitor; some per-call monitors. The channel-level monitor will destroy all the channels upon network error. The workaround I mentioned above is a one-line change in the call-level monitor, which cancels the call upon network error. 

But I definitely agree that this workaround is not good enough for long-term use. Since the customer is already using some other workaround, I think fixing #14292 is the right choice. That will be much cleaner.

**Comment by mikelehen** [comment]
FYI- In case there's any confusion, the customer is not calling grpc_channel_destroy() or "using some other workaround" as far as I know.

The repro project linked in the issue description is an iOS app using the Firebase iOS SDK for Cloud Firestore (which uses grpc internally).  I work on the SDK, and all we we do is initiate a streaming RPC request to the backend. The user-facing issue is that upon a network change (e.g. switching from wifi to cellular), it takes ~90 seconds for the streaming RPC to fail (at which point we can retry it and resume connectivity with the backend). This timeout is too long.

I haven't followed all of the discussion above but I just wanted to clear up the use case / symptom in case there was any confusion leading to false assumptions.

**Comment by AspirinSJL** [comment]
@mikelehen Thanks for clarifying!

`grpc_channel_destroy()` is not called by the customer explicitly; it's called in gRPC Objective-C when some network error happens.

Sorry for misunderstanding the workaround status! The customer workaround that we thought is for this issue is actually for another issue. Then I think the workaround I proposed can be used if we want a quick fix. Fixing #14292 is a better solution to this issue, but it will be a much larger change and will take longer to finish.

**Comment by AspirinSJL** [comment]
After discussion, we decided to cancel the call in Objective-C layer when the per-call monitor detects network error. Re-assigned to @muxi .
